connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java

```diff
@@ -0,0 +1,235 @@
+            if (!result.success()) {
+                markAsFailed();
+                errorHandlingMetrics.recordSkipped();
+            }
+            context.result(result);
+            return result;
+        } catch (Exception e) {
+            errorHandlingMetrics.recordFailure();
+            markAsFailed();
+
+            Result<V> exResult = new Result<>(e);
+            context.result(exResult);
+
+            if (!tolerated.isAssignableFrom(e.getClass())) {
+                context.result(new Result<>(e));
+                throw new ConnectException("Unhandled exception in error handler", e);
+            }
+
+            if (!withinToleranceLimits()) {
+                throw new ConnectException("Tolerance exceeded in error handler", e);
+            }
+
+            errorHandlingMetrics.recordSkipped();
+            return exResult;
+        }
+    }
+
+    // Visible for testing
+    void markAsFailed() {
+        errorHandlingMetrics.recordErrorTimestamp();
+        totalFailures++;
+    }
+
+    // Visible for testing
+    boolean withinToleranceLimits() {
+        switch (config.toleranceLimit()) {
+            case NONE:
+                if (totalFailures > 0) return false;
+            case ALL:
+                return true;
+            default:
+                throw new ConfigException("Unknown tolerance type: {}", config.toleranceLimit());
+        }
+    }
+
+    // Visible for testing
+    boolean checkRetry(long startTime) {
+        return (time.milliseconds() - startTime) < config.retryTimeout();
+    }
+
+    // Visible for testing
+    void backoff(int attempt, long deadline) {
+        int numRetry = attempt - 1;
+        long delay = RETRIES_DELAY_MIN_MS << numRetry;
+        if (delay > config.retryDelayMax()) {
+            delay = ThreadLocalRandom.current().nextLong(config.retryDelayMax());
+        }
+        if (delay + time.milliseconds() > deadline) {
+            delay = deadline - time.milliseconds();
+        }
+        log.debug("Sleeping for {} millis", delay);
+        time.sleep(delay);
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        config = new RetryWithToleranceExecutorConfig(configs);
+    }
+
+    public void setMetrics(ErrorHandlingMetrics errorHandlingMetrics) {
+        this.errorHandlingMetrics = errorHandlingMetrics;
+    }
+
+    static class RetryWithToleranceExecutorConfig extends AbstractConfig {
+        public RetryWithToleranceExecutorConfig(Map<?, ?> originals) {
+            super(getConfigDef(), originals, true);
+        }
+
+        public long retryTimeout() {
+            return getLong(RETRY_TIMEOUT);
+        }
+
+        public long retryDelayMax() {
+            return getLong(RETRY_DELAY_MAX_MS);
+        }
+
+        public ToleranceType toleranceLimit() {
+            return "ALL".equals(getString(TOLERANCE_LIMIT).toUpperCase(Locale.ROOT)) ? ToleranceType.ALL
+                    : ToleranceType.NONE;
+        }
```

Comment:
How about moving the parsing of the string into a `ToleranceType` into the enum itself as a static method?